### ISSUE TITLE
IPLD + Search

### ISSUE THREAD

**Issue by vmx** [ipld/ipld#43]
This issue summarizes what was discussed at the [IPFS Developer Meetings 2018] in Berlin at a discussion seesion about IPLD + Search.

Please use this issues for all things related to IPLD + Search. Feel free to go crazy!

Use cases
---------

I’d like to ask the attendees to post their uses cases as comment. Especially from @b5, @jonnycrunch and @bigs.

Protoype proposal
-----------------

Build up an index that is separate from the DAG and local to the peer. For a prototype [Noise] will be used. As Noise only supports JSON, you need to have a JSON compatible representation of the object that is stored in IPLD. For [js-ipld] this is a call to `resolve()` with the root `/` as path.

Noise is an embedded search index written in Rust with a Node.js API. There is no proper C-API yet. It's WIP (see https://github.com/pipedown/noise/pull/63), help/comments are appreciated. Until then, the easiest way is to use it through a HTTP wrapper. There is one for storing data (https://github.com/vmx/noise-dataload) and one for serving up the data (https://github.com/vmx/noise-indexserve). More details on how to use them can be found at my [Exploring data with Noise] blog post.

Next steps
----------

 - Have people play with Noise
 - Get Noise C-API done
 - Write a Go binding to the C-API (as most people seem to use go-ipfs and not js-ipfs).

[IPFS Developer Meetings 2018]: https://github.com/ipfs/developer-meetings
[Noise]: https://github.com/pipedown/noise/
[js-ipld]: https://github.com/ipld/js-ipld
[Exploring data with Noise]: https://vmx.cx/cgi-bin/blog/index.cgi/exploring-data-with-noise%3A2017-12-12%3Aen%2CNoise%2CNode%2CJavaScript%2CRust

**Comment by gedw99** [comment]
Search everything so that you can have a Google drive like GUI and find anything in it after years and years.

OCR would be part of it.

Should be written in golang because search is a core thing. Don't want a nodejs system running in backend 

Facetted based search is very powerful because you can also build powerful drill down based search.


Distributed  storage on the index.
Distributed search.
Bleve does this using alias, a very simple and extensible approach.



Bleve fulfils all these needs.

**Comment by vmx** [comment]
@gedw99 You are right, there's currently only a Node.js binding for Noise. As soon as Noise exposes a C API, there can easily be a Go binding for it.

I agree that having something in Go is nice for go-ipfs, but there's also other implementations like js-ipfs. Those other implementations would then have a similar problem with Bleve.

**Comment by gedw99** [comment]
I understand.

Just expose bleve over the JSON API.
That's what it's there for. To allow easy interoperability.
That's what blast ( that wraps bleve ) also did , enabling storage and
search to both be fully distributed and callable via a JSON interface.

I am a very broken record :)

**Comment by lgierth** [comment]
How does this make use of IPLD/IPFS? Is the noise index stored in IPFS and the "client" part understands how to load it? Can the index be sharded, and only those parts be loaded that are needed for a query?

**Comment by lgierth** [comment]
Ah, so it sounds like you're thinking more of making an IPFS node's data searchable. The angle from which I was thinking about search is efficient sharded indexes for search in e.g. websites (docs.ipfs.io, wikipedia, etc), so I was confused at first.

**Comment by gedw99** [comment]
hey @lgierth 
yep you got it exactly, make the IPFS data searchable - thats what i was meaning from the start.

You get multilanguage search ( because bleve has most of the worlds stemmers, which is a ton of work btw ).
You get facets which is incredibly useful stuff.

**Comment by vmx** [comment]
@edsilv It would be cool if you could also provide your use case as a comment on this issue.

**Comment by edsilv** [comment]
Hey @vmx - here's the IIIF Discovery group's work so far: https://github.com/IIIF/discovery/issues

**Comment by vmx** [comment]
Here's another use case (from @ibnesayeed). He stores Mementos which link to some namespace. As it's a DAG, the namespace can't easily link back to the Mementos, hence it's difficult to get all the Mementos that link to a namespace.

With search you could query for all Mementos linking to a certain namespace.

I know that there's also other solutions to the problem, but search could be one.
